### PR TITLE
feat: add initial implementation of neighbourhood logic

### PR DESCRIPTION
**Pull request by simonusher**

### CODE REVIEW

File: srai/neighbourhood/neighbourhood.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return neighbours_at_distance
+
+    def _get_neighbours_with_distances(
+        self, index: IndexType, distance: int
+    ) -> Set[Tuple[IndexType, int]]:
+        visited = set()
+        visited_with_distances = set()
+        to_visit: Queue[Tuple[IndexType, int]] = Queue()
+
+        visited.add(index)
```
Comment (piotrgramacki) [inline srai/neighbourhood/neighbourhood.py:77]: Why add this first region twice? Here and in `while` loop?

File: srai/neighbourhood/neighbourhood.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+class Neighbourhood(ABC, Generic[IndexType]):
+    """Neighbourhood interface."""
+
+    @abstractmethod
+    def get_neighbours(self, index: IndexType) -> Set[IndexType]:
+        """
+        Get the direct neighbours of a region using its index.
+
+        Args:
+            index (Any): Unique identifier of the region.
```
[the remark this answers is not in the document]
Comment (simonusher) [reply srai/neighbourhood/neighbourhood.py]: You may be right. I changed that

File: srai/neighbourhood/neighbourhood.py (hunk 1 above)
Comment (simonusher) [reply srai/neighbourhood/neighbourhood.py]: I changed it a couple of times and ended up with this. I simplified it, so I hope it now looks better.

File: srai/neighbourhoods/neighbourhood.py [hunk 3]
```diff
@@ -10,51 +10,63 @@
...
 
         Args:
-            index (Any): Unique identifier of the region.
+            index (IndexType): Unique identifier of the region.
                 Dependant on the implementation.
             distance (int): Maximum distance to the neighbours.
 
         Returns:
-            List[Any]: Indexes of the neighbours.
+            List[IndexType]: Indexes of the neighbours.
```
Comment (piotrgramacki) [inline srai/neighbourhoods/neighbourhood.py:52]: `List` or `Set`? Now it is inconsistent with typings

**Review by piotrgramacki — DISMISSED** [review]

File: srai/neighbourhood/neighbourhood.py (hunk 1 above)
Comment (piotrgramacki) [reply srai/neighbourhood/neighbourhood.py]: Thanks!

File: srai/neighbourhood/neighbourhood.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (piotrgramacki) [reply srai/neighbourhood/neighbourhood.py]: Thanks!

**Review by piotrgramacki — DISMISSED** [review]

File: srai/neighbourhoods/neighbourhood.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+            visited.add(current)
+            visited_with_distances.add((current, current_distance))
+            if current_distance < distance:
+                current_neighbours = self.get_neighbours(current)
+                for neighbour in current_neighbours:
+                    if neighbour not in visited:
+                        to_visit.put((neighbour, current_distance + 1))
+
+        return visited_with_distances
```
Comment (RaczeQ) [inline srai/neighbourhoods/neighbourhood.py:101]: wouldn't it be better to replace it with a `dict`?
```suggestion    
        visited_indexes = dict()    
        to_visit: Queue[Tuple[IndexType, int]] = Queue()

        to_visit.put((index, 0))

        while not to_visit.empty():
            current_index, current_distance = to_visit.get()

            visited_indexes[current_index] = current_distance
            if current_distance < distance:
                current_neighbours = self.get_neighbours(current_index)
                for neighbour in current_neighbours:
                    if neighbour not in visited_indexes:
                        to_visit.put((neighbour, current_distance + 1))

        return set(visited_indexes.items())
```

**Review by RaczeQ — APPROVED** [review]

**Pull request merged**